airflow/api_connexion/endpoints/user_endpoint.py

```diff
@@ -124,9 +124,21 @@ def patch_user(username, update_mask=None):
     if user is None:
         detail = f"The User with username `{username}` was not found"
         raise NotFound(title="User not found", detail=detail)
-
-    # Get fields to update. 'username' is always excluded (and it's an error to
-    # include it in update_maek).
+    # Check unique username
+    new_username = data.get('username')
+    if new_username:
+        usr = security_manager.find_user(username=new_username)
+        if usr and usr != user:
+            raise AlreadyExists(detail=f"The username `{new_username}` already exists")
+
+    # Check unique email
+    email = data.get('email')
+    if email:
+        usr = security_manager.find_user(email=email)
+        if usr and usr != user:
+            raise AlreadyExists(detail=f"The email `{email}` already exists")
```

Comment:
I think we could save a round-trip to the database in many cases with

```python
if email and email != user.email:
```

Also, could either `user` and `email` be empty strings in the data schema? We should emit 400 for those instead of treating them like missing fields (as the logic implemented here would do).

Reply:
I think instead of `required` we should set `minLength: 1` instead? This would allow either field to be missing (implying unchanged), but if it's present, it must not be empty.

Reply:
I think one important consideration is that those fields are pre-filled in the web UI (I think by Flask-Appbuilder?) so setting them as required imposes no cost to the user, but that's not the case for the API. The PATCH verb is designed to allow the user to not send fields they don't want to modify, so requiring fields is kind of against the spirit IMO, especially for `email` (`username` is already needed to access the endpoint anyway so requiring it in the payload is only unnecessary duplication and slightly less problematic).

But even if we decide to make these fields required, they still need to have `minLength: 1` IMO. It's very non-obvious that setting those fields to an empty string implies not changing them. The expected behaviour for providing an empty string in a field is setting that field to an empty string.

Reply:
Probably we should have a different form for update? Post form should have some fields as required

Reply:
Hmm, that's not what the JSON Schema specification says (OpenAPI takes both `required` and `minLength` directly from it), so this is more likely a bug in the implementation. In that case, maybe we should leave both fields out from the specification, but mention that both must be non-empty if supplied, and validate them manually instead.

Reply:
I would suggest we put these as required values at the expense of user experience. Now, the post endpoint thinks that username/email are required and doesn't consider first_name/last_name as required values. This means that we can create users without first_name and last_name. The schema for users doesn't have required fields, the same as the API spec. I'm proposing to make first_name, last_name, username, and email required fields.

The patch endpoint is intended to be used to edit other users, not only the user making the request. Therefore making the fields required may not actually be at the expense of user experience but a normal thing

Reply:
Let's set `minLength` then (which implies `required` and is the strictest option).